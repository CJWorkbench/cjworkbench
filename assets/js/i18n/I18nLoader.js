import React from 'react'
import { I18nProvider } from '@lingui/react'
import fetchCatalog from './catalogs'

export class I18nLoader extends React.Component {
  state = {
    catalogs: {},
  }
  
  loadCatalog = (language) => {
      this.setState(state => ({
          catalogs: {
            ...state.catalogs,
            [language]: fetchCatalog(language)
          }
      }))
  }
  
  componentDidMount() {
    this.loadCatalog(this.props.language)
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { language } = nextProps
    const { catalogs } = nextState

    if (language !== this.props.language && !catalogs[language]) {
      this.loadCatalog(language)
      return false
    }

    return true
  }

  render () {
    const { children, language } = this.props
    const { catalogs } = this.state

    // Skip rendering when catalog isn't loaded.
    if (!catalogs[language]) {
        return null
    }

    return (
      <I18nProvider language={language} catalogs={catalogs}>
        {children}
      </I18nProvider>
    )
  }
}
