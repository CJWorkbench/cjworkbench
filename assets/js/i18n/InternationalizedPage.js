import React from 'react'
import PropTypes from 'prop-types'
import { en, el } from 'make-plural/plurals'
import { i18n } from '@lingui/core'
import { I18nProvider } from '@lingui/react'
import { currentLocaleId } from './locales'
import Spinner from '../Spinner'

__webpack_public_path__ = window.STATIC_URL + 'bundles/' // eslint-disable-line

async function loadLocale (localeId) {
  const { messages } = await import(
    /* webpackChunkName: "[request]" */
    `@lingui/loader!../../locale/${localeId}/messages.po`
  )
  i18n.loadLocaleData({
    en: { plurals: en },
    el: { plurals: el }
  })
  i18n.load(localeId, messages)
  i18n.activate(localeId)
  return i18n
}

function useActivatedI18n (localeId) {
  const [i18n, setI18n] = React.useState(null)

  React.useEffect(() => {
    loadLocale(currentLocaleId).then(setI18n)
  }, [])

  return i18n
}

export default function InternationalizedPage (props) {
  const { children } = props
  const i18n = useActivatedI18n(currentLocaleId)

  if (i18n === null) {
    return <Spinner />
  }

  return (
    <I18nProvider i18n={i18n} forceRenderOnLocaleChange={false}>
      {children}
    </I18nProvider>
  )
}
InternationalizedPage.propTypes = {
  children: PropTypes.node.isRequired
}
