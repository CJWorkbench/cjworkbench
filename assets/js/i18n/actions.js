const SET_LOCALE = 'SET_LOCALE';

function reduceSetLocale(state, action){
  return {
    ...state,
    locale: action.locale
  };
}

export function setLocaleAction(locale){
  return {
    type: SET_LOCALE,
    locale: locale
  }
}

export const reducerFunctions = {
  [SET_LOCALE]: reduceSetLocale
}

export function localeReducer(state, action){
  if (action.type in reducerFunctions) {
    // Run a registered reducer
    return reducerFunctions[action.type](state, action)
  }
  
  return state;
}
