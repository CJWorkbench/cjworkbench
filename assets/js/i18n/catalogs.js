import enCatalog from '../../locale/en/messages.po'
import elCatalog from '../../locale/el/messages.po'

const catalogs = {
  en: enCatalog,
  el: elCatalog
}

export default function fetchCatalog (language) {
  return catalogs[language]
}
