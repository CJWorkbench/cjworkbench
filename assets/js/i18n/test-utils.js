import { I18nProvider } from '@lingui/react'
import React from 'react'
import { shape, object } from 'prop-types'
import { mount, shallow } from 'enzyme'

/*
 * Below lie utils for testing with i18n.
 * See https://lingui.js.org/guides/testing.html
 */
// Create the I18nProvider to retrieve context for wrapping around.
const intlProvider = new I18nProvider({
  language: 'en',
  catalogs: {
    en: {}
  }
}, {})

const {
  linguiPublisher: {
    i18n: originalI18n
  }
} = intlProvider.getChildContext()

// You customize the i18n object here:
export const i18n = {
  ...originalI18n,
  // provide _ macro, for trying to unwrap default/id or just passing down the key
  _: key => {
    if (typeof key.defaults === 'string') return key.defaults
    if (typeof key.id === 'string') return key.id
    return key
  }
}

/**
 * When using Lingui `withI18n` on components, props.i18n is required.
 */
function nodeWithI18nProp (node) {
  return React.cloneElement(node, { i18n })
}

/*
 * Methods to use
 */
export function shallowWithI18n (node, { context } = {}) {
  return shallow(
    nodeWithI18nProp(node),
    {
      context: Object.assign({}, context, { i18n })
    }
  )
}

export function mountWithI18n (node, { context, childContextTypes } = {}) {
  const newContext = Object.assign({}, context, { linguiPublisher: { i18n } })
  /*
   * I18nProvider sets the linguiPublisher in the context for withI18n to get
   * the i18n object from.
   */
  const newChildContextTypes = Object.assign({},
    {
      linguiPublisher: shape({
        i18n: object.isRequired
      }).isRequired
    },
    childContextTypes
  )
  return mount(
    nodeWithI18nProp(node),
    {
      context: newContext,
      childContextTypes: newChildContextTypes
    }
  )
}
