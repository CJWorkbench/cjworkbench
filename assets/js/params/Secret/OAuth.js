import React from 'react'
import PropTypes from 'prop-types'
import { Trans } from '@lingui/macro'

export default class OAuth extends React.PureComponent {
  static propTypes = {
    name: PropTypes.string.isRequired,
    secretMetadata: PropTypes.shape({
      name: PropTypes.string.isRequired
    }), // null if not set
    startCreateSecret: PropTypes.func.isRequired, // func(name) => undefined
    deleteSecret: PropTypes.func.isRequired, // func(name) => undefined
    secretLogic: PropTypes.shape({
      provider: PropTypes.oneOf(['oauth1a', 'oauth2']),
      service: PropTypes.oneOf(['google', 'intercom', 'twitter'])
    })
  }

  handleClickConnect = () => {
    const { name, startCreateSecret } = this.props
    startCreateSecret(name)
  }

  handleClickDisconnect = () => {
    const { name, deleteSecret } = this.props
    deleteSecret(name)
  }

  render () {
    const { secretMetadata } = this.props

    let contents
    if (secretMetadata) {
      contents = (
        <>
          <p className='secret-name'>{secretMetadata.name}</p>
          <button type='button' className='disconnect' onClick={this.handleClickDisconnect}><Trans id='js.params.Secret.OAuth.signOut.button'>Sign out</Trans></button>
        </>
      )
    } else {
      contents = (
        <button type='button' className='connect' onClick={this.handleClickConnect}><Trans id='js.params.Secret.OAuth.connectAccount.button'>Connect account</Trans></button>
      )
    }

    return (
      <div className='oauth-connect-parameter'>
        {contents}
      </div>
    )
  }
}
