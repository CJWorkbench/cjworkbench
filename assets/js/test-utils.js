/* globals jest */
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import promiseMiddleware from 'redux-promise-middleware'
import errorMiddleware from './error-middleware'
import { workflowReducer } from './workflow-reducer'

// Returns new mock function that returns given json. Used for mocking "get" API calls
export function jsonResponseMock (json) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve(json)
  )
}

// Returns new mock function that gives an OK HTTP response. Use for mocking "set" API calls
export function okResponseMock () {
  return jsonResponseMock(null)
}

// Helper function to return a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...).then(...)
// Technically: this is designed to resolve on the next macrotask
// https://stackoverflow.com/questions/25915634/difference-between-microtask-and-macrotask-within-an-event-loop-context
export function tick () {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function sleep (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

/**
 * Return `[ notify, promise ]`; `promise` completes when `notify` is called.
 *
 * Usage:
 *
 *     const [ notify, condition ] = createConditionVariable
 *     const done = condition.then(do_something)
 *     await Promise.all([sleep(0.1).then(notify), done])
 */
export function createConditionVariable () {
  let notify
  const promise = new Promise(resolve => { notify = resolve })
  return [notify, promise]
}

/**
 * Mock Redux store with in-production reducers.
 *
 * This is for integration-style tests: they test the reducer _and_ the caller
 * at the same time.
 *
 * Usage:
 *
 *     const api = { undo: jest.fn() }
 *     const store = mockStoreWithReducer({'workflow': ..., ...}, api)
 *     store.dispatch(Actions.undo())
 *     expect(api.undo).toHaveBeenCalled()
 *     expect(store.getState().workflow).toEqual(...)
 */
export function mockStore (initialState, api = {}) {
  const middlewares = [errorMiddleware(), promiseMiddleware, thunk.withExtraArgument(api)]
  const store = createStore(workflowReducer, initialState, applyMiddleware(...middlewares))
  return store
}
