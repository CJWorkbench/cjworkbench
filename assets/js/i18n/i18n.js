import { setupI18n } from "@lingui/core"
import { currentLocale } from './locales'
import { fetchCatalog } from './catalogs'

export default const i18n = setupI18n({
   language: currentLocale,
   catalogs: {
       currentLocale: fetchCatalog(currentLocale)
   }
})
