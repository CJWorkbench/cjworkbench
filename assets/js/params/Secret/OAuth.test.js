/* globals describe, expect, it, jest */
import React from 'react'
import OAuth from './OAuth'
import { mountWithI18n } from '../../i18n/test-utils'

describe('OAuth', () => {
  const wrapper = (extraProps) => {
    return mountWithI18n(
      <OAuth
        name='x'
        startCreateSecret={jest.fn()}
        deleteSecret={jest.fn()}
        secretMetadata={{ name: 'a secret' }}
        secretLogic={{ service: 'google' }}
        isOwner
        {...extraProps}
      />
    )
  }

  it('matches snapshot', () => {
    const w = wrapper({})
    expect(w).toMatchSnapshot()
  })

  it('does not show "Sign in" link when not owner', () => {
    const w = wrapper({ secretMetadata: null, isOwner: false })
    expect(w.find('button.connect')).toHaveLength(0)
  })

  it('renders without a secret', () => {
    const w = wrapper({ secretMetadata: null })
    expect(w.find('button.connect')).toHaveLength(1)
    w.find('button.connect').simulate('click')
    expect(w.prop('startCreateSecret')).toHaveBeenCalledWith('x')
  })

  it('does not show "Sign out" link when not owner', () => {
    const w = wrapper({ isOwner: false })
    expect(w.find('button.disconnect')).toHaveLength(0)
  })

  it('disconnects', () => {
    const w = wrapper({ secretMetadata: { name: 'foo@example.org' } })
    w.find('button.disconnect').simulate('click')
    expect(w.prop('deleteSecret')).toHaveBeenCalledWith('x')
  })
})
