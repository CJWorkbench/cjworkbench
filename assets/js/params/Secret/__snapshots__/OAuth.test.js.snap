// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OAuth matches snapshot 1`] = `
<OAuth
  deleteSecret={[MockFunction]}
  name="x"
  secretLogic={
    Object {
      "service": "google",
    }
  }
  secretMetadata={
    Object {
      "name": "a secret",
    }
  }
  startCreateSecret={[MockFunction]}
>
  <div
    className="oauth-connect-parameter"
  >
    <p
      className="secret-name"
    >
      a secret
    </p>
    <button
      className="disconnect"
      onClick={[Function]}
      type="button"
    >
      <WithI18n
        defaults="Sign out"
        id="js.params.Secret.OAuth.signOut.button"
      >
        <I18n
          update={true}
          withHash={true}
        >
          <Trans
            defaults="Sign out"
            id="js.params.Secret.OAuth.signOut.button"
          >
            <Render
              value="js.params.Secret.OAuth.signOut.button"
            >
              js.params.Secret.OAuth.signOut.button
            </Render>
          </Trans>
        </I18n>
      </WithI18n>
    </button>
  </div>
</OAuth>
`;
